server/src/main/java/org/elasticsearch/action/bulk/WriteMemoryLimits.java

```diff
@@ -20,17 +20,50 @@
 package org.elasticsearch.action.bulk;
 
 import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
 
 import java.util.concurrent.atomic.AtomicLong;
 
 public class WriteMemoryLimits {
 
+    // TODO: Adjust
+    public static final Setting<ByteSizeValue> MAX_INDEXING_BYTES =
+        Setting.memorySizeSetting("indices.write.limit", "10%", Setting.Property.NodeScope, Setting.Property.Dynamic);
+
     private final AtomicLong writeBytes = new AtomicLong(0);
     private final AtomicLong replicaWriteBytes = new AtomicLong(0);
+    private volatile long writeLimits;
+
+    public WriteMemoryLimits(Settings settings, ClusterSettings clusterSettings) {
+        this.writeLimits = MAX_INDEXING_BYTES.get(settings).getBytes();
+        clusterSettings.addSettingsUpdateConsumer(MAX_INDEXING_BYTES, value -> writeLimits = value.getBytes());
+    }
 
     public Releasable markWriteOperationStarted(long bytes) {
-        writeBytes.addAndGet(bytes);
-        return () -> writeBytes.getAndAdd(-bytes);
+        return markWriteOperationStarted(bytes, false);
+    }
+
+    public Releasable markWriteOperationStarted(long bytes, boolean forceExecution) {
+        long writeBytes = this.writeBytes.addAndGet(bytes);
+        long replicaWriteBytes = this.replicaWriteBytes.get();
+        long totalBytes = writeBytes + replicaWriteBytes;
+        long localWriteLimits = this.writeLimits;
+        if (forceExecution == false && totalBytes > localWriteLimits) {
```

Comment:
I discussed this with Henning, and we think that a second replica limit would be useful here (e.g. defaulting to 1,5 times the `indices.write.limit`). In the past we have been rejecting replica operations at the real-memory circuit breaker limit (i.e. 95%), which can have adverse effects on the rest of the system. We should have another limit for replicas much lower than that. It won't solve all problems with handling primary vs replica bytes, but it puts an additional safeguard into the system.